File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    term = {}
    
    for t in terms:
        ts = t.split()
        term[ts[0]] = int(ts[1])
    
    for i, p in enumerate(privacies):
        date, x = p.split()
        year = int(date[0:4])
        month = int(date[5:7]) + term[x]
        day = int(date[8:]) - 1
        
        if day == 0:
            month -= 1
            day = 28
        if month > 12:
            if month % 12 == 0:
                year += (month // 12) - 1
                month = 12
            else:
                year += month // 12
                month %= 12

        print(year, month, day)
        tdy = list(map(int, today.split('.')))
        
        if tdy[0] > year:
            answer.append(i+1)
            continue
        elif tdy[1] > month and tdy[0] == year:
            answer.append(i+1)
            continue
        elif tdy[2] > day and tdy[1] == month and tdy[0] == year:
            answer.append(i+1)
            continue
    
    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        ArrayList<Integer> answer = new ArrayList<>();
        int t_year = 0;
        int t_month = 0;
        int t_day = 0;
        Map<String, Integer> map = new HashMap<>();
        
        String[] tdy = today.split("\\.");
        t_year = Integer.parseInt(tdy[0]);
        t_month = Integer.parseInt(tdy[1]);
        t_day = Integer.parseInt(tdy[2]);
        
        
        for(int i = 0; i < terms.length; i++) {
            String t[] = terms[i].split(" ");
            map.put(t[0], Integer.parseInt(t[1]));
        }
        
        
        
        for(int i = 0; i < privacies.length; i++) {
            String priv[] = privacies[i].split(" ");
            int tm = map.get(priv[1]);
            String[] date = priv[0].split("\\.");
            int year = Integer.parseInt(date[0]);
            int month = Integer.parseInt(date[1]);
            int day = Integer.parseInt(date[2]);
            
            year += tm / 12;
            month += tm % 12;
            
            // 12월이 넘어갈 경우 처리
            if(month > 12) {
                year += month / 12;
                month = month % 12;
            }
            
            day -= 1;
            if(day == 0) {
                day = 28;
                month -= 1;
            }
            
            // month가 0일 경우 처리
            if(month == 0) {
                year -= 1;
                month = 12;
            }
            
            
            
            if(t_year > year) {
                answer.add(i + 1);
                continue;
            } 
            else if(t_year < year) {
                continue;
            }
            else {
                if(t_month > month) {
                    answer.add(i + 1);
                    continue;
                }
                else if(t_month < month) {
                    continue;
                }
                else {
                    if(t_day > day) {
                        answer.add(i + 1);
                        continue;
                    }
                    else  {
                        continue;
                    }
                }
            }
            
        }
        
        int[] result = new int[answer.size()];
        for(int i = 0; i < answer.size(); i++) {
            result[i] = answer.get(i);
        }

        return result;
    }
}